cmd/clusterctl/client/cluster/objectgraph.go

```diff
@@ -615,14 +669,46 @@ func (o *objectGraph) setSoftOwnership() {
 	}
 
 	clusterClasses := o.getClusterClasses()
+
+	// Collect all cross-ns ClusterClasses to exclude from deletion
+	for _, clusterClass := range clusterClasses {
+		for _, cluster := range o.externalObjects {
+			if cluster.identity.GroupVersionKind().GroupKind() != clusterv1.GroupVersion.WithKind("Cluster").GroupKind() {
+				continue
+			}
+
+			// if the cluster uses a managed topology and uses the clusterclass
+			// check the namespace of the class
+			className, hasName := cluster.additionalInfo[clusterTopologyNameKey]
+			classNamespace, hasNamespace := cluster.additionalInfo[clusterTopologyNamespaceKey]
+			if namespace, ok := classNamespace.(string); !hasNamespace || !ok {
+				classNamespace = cmp.Or(namespace, cluster.identity.Namespace)
+			}
+
+			// Set shouldNotDelete if clusterClass already referenced by some external cluster
+			if hasName && className == clusterClass.identity.Name && clusterClass.identity.Namespace == classNamespace {
+				clusterClass.shouldNotDelete = true
+			}
+		}
+	}
+
 	// Cluster that uses a ClusterClass are soft owned by that ClusterClass.
 	for _, clusterClass := range clusterClasses {
 		for _, cluster := range clusters {
 			// if the cluster uses a managed topology and uses the clusterclass
 			// set the clusterclass as a soft owner of the cluster.
-			if className, ok := cluster.additionalInfo[clusterTopologyNameKey]; ok {
-				if className == clusterClass.identity.Name && clusterClass.identity.Namespace == cluster.identity.Namespace {
-					cluster.addSoftOwner(clusterClass)
+			className, hasName := cluster.additionalInfo[clusterTopologyNameKey]
+			classNamespace, hasNamespace := cluster.additionalInfo[clusterTopologyNamespaceKey]
+			if namespace, ok := classNamespace.(string); !hasNamespace || !ok {
+				classNamespace = cmp.Or(namespace, cluster.identity.Namespace)
+			}
+
+			if hasName && className == clusterClass.identity.Name && clusterClass.identity.Namespace == classNamespace {
+				cluster.addSoftOwner(clusterClass)
+
+				// Set shouldNotDelete if the referenced clusterClass is in a different namespace
+				if classNamespace != cluster.identity.Namespace {
```

Comment:
Might be I'm missing something, but if I'm not wrong, we are not reading CC that exists in other NS

If correct, this might increase the complexity of the PR quite a lot, because we have to figure it out also  how to read templates referenced in the CC from other ns, and we do not now know template kinds upfront.

Also, it will add a new entire dimension - objects from other ns - to our test matrix

If this can help, I'm available to talk about this in person and help to figure out a way forward